doc/SAI-Proposal-ARS.md

```diff
@@ -0,0 +1,1327 @@
+
+ARS attempts to address some of the shortcomings in the existing network deployments.
+
+## 2.0 Existing Forwarding Decision Model
+Today the routing protocol or SDN controller decides the reachability of a given destination and finds all the possible paths. These paths may be equal cost or unequal cost. Decision to choose one of the paths out of all available paths is done in the switch pipeline based on a computed hash. This path selection is static in nature based on the packet fields. Hash based selection doesn't take into account the dynamic state of the local or end to end path.
+
+There are control plane protocols for traffic engineering a path but involves a control plane decision that is very slow to react to changing traffic patterns in the network.
+
+Some of the observed shortcomings are
+- Load balancing requires different sizes of flows to be evenly distributed across aggregate member links.
+- Size and duration of flows is not considered when performing aggregate member assignment.
+- Once a member link is assigned there is no consideration of degradation of the link because of overloading or queue congestion.
+- Polarization: where traffic takes a particular path even if the path is congested.
+- Local Link Quality: There is no concept of link quality in path selection.
+- End to End Path Quality: There is no concept of path quality between two endpoints.
+- Link failure: where control plane rebalances the available paths but results in blackholing of traffic while rebalance is going on.
+
+These shortcomings result in blackholing, slow convergence, dropped traffic and/or under utilization of the network thereby resulting in under utilization of expensive accelerators.
+
+## 3.0 Adaptive Routing and Switching
+
+ARS attempts to address these shortcomings by
+1. Accounting for local and remote link quality for path selection
+2. Use quality metrics to account for a changing traffic pattern in the network
+3. Dynamic path assignment and reassignment without any interaction with the control plane
+4. Minimize traffic drop when link flaps by HW based link down detection and path reassignment
+
+### 3.1 Terms and Acronyms
+| Acronym   | Term      |
+|:----------|:----------|
+| Micro Flow    | A single traffic flow typically identified by 5 tuple    |
+| Macro Flow    | collection of micro flows that hash to same value   |
+| Idle Time | Time when there are no packets observed within a macro flow |
+| Flowlet    | contagious subsequence of packets within a macro flow separated by an idle time interval    |
+| Macro Flow Table    | HW table to store per macro flow state    |
+| EWMA | Exponential weighted moving average |
+| ARS    | Adaptive routing and switching: Path selection is influenced by quality metrics    |
+| NHG | Next Hop Group |
+
+### 3.2 Key Idea behind ARS
+1. ARS is flowlet based
+
+      - *A flow is viewed as a sequence of flowlets separated by idle time*

```

Comment:
Nit: Should the "flow" in this section be replaced by "Macro Flow"? Traditionally, we think of micro-flows as flows, so explicit usage of micro-flow or macro-flow in the rest of this document will help avoid misunderstanding.

Reply:
Flow and Macro flow is used interchangeably and given context should help. I have opened a new PR and would avoid churn in that if possible.